filebeat/fileset/fileset.go

```diff
@@ -0,0 +1,257 @@
+	}
+
+	pipeline_id, err := fs.getPipelineID()
+	if err != nil {
+		return err
+	}
+	fs.vars["beat"] = map[string]interface{}{
+		"pipeline_id": pipeline_id,
+	}
+
+	return nil
+}
+
+// manifest structure is the representation of the manifest.yml file from the
+// fileset.
+type manifest struct {
+	ModuleVersion  string                   `config:"module_version"`
+	Vars           []map[string]interface{} `config:"var"`
+	IngestPipeline string                   `config:"ingest_pipeline"`
+	Prospector     string                   `config:"prospector"`
+}
+
+// readManifest reads the manifest file of the fileset.
+func (fs *Fileset) readManifest() (*manifest, error) {
+	cfg, err := common.LoadFile(filepath.Join(fs.modulePath, fs.name, "manifest.yml"))
+	if err != nil {
+		return nil, fmt.Errorf("Error reading manifest file: %v", err)
+	}
+	var manifest manifest
+	err = cfg.Unpack(&manifest)
+	if err != nil {
+		return nil, fmt.Errorf("Error unpacking manifest: %v", err)
+	}
+	return &manifest, nil
+}
+
+// evaluateVars resolves the fileset variables.
+func (fs *Fileset) evaluateVars() (map[string]interface{}, error) {
+	var err error
+	vars := map[string]interface{}{}
+	vars["builtin"], err = fs.getBuiltinVars()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, vals := range fs.manifest.Vars {
+		var exists bool
+		name, exists := vals["name"].(string)
+		if !exists {
+			return nil, fmt.Errorf("Variable doesn't have a string 'name' key")
+		}
+
+		value, exists := vals["default"]
+		if !exists {
+			return nil, fmt.Errorf("Variable %s doesn't have a 'default' key", name)
+		}
+
+		// evaluate OS specific vars
+		osVals, exists := vals["os"].(map[string]interface{})
+		if exists {
+			osVal, exists := osVals[runtime.GOOS]
+			if exists {
+				value = osVal
+			}
+		}
+
+		vars[name], err = resolveVariable(vars, value)
+		if err != nil {
+			return nil, fmt.Errorf("Error resolving variables on %s: %v", name, err)
+		}
+	}
+
+	// overrides from the config
+	for name, val := range fs.fcfg.Var {
+		vars[name] = val
+	}
+
+	return vars, nil
+}
+
+// resolveVariable considers the value as a template so it can refer to built-in variables
+// as well as other variables defined before them.
+func resolveVariable(vars map[string]interface{}, value interface{}) (interface{}, error) {
+	switch v := value.(type) {
+	case string:
+		return applyTemplate(vars, v)
+	case []interface{}:
+		transformed := []interface{}{}
+		for _, val := range v {
+			s, ok := val.(string)
+			if ok {
+				transf, err := applyTemplate(vars, s)
+				if err != nil {
+					return nil, fmt.Errorf("array: %v", err)
+				}
+				transformed = append(transformed, transf)
+			} else {
+				transformed = append(transformed, val)
+			}
+		}
+		return transformed, nil
+	}
+	return value, nil
+}
+
+// applyTemplate applies a Golang text/template
+func applyTemplate(vars map[string]interface{}, templateString string) (string, error) {
+	tpl, err := template.New("test").Parse(templateString)
```

Comment:
I think that doesn't really matter, but also "test" is not really good since this is not a test :). I'll make it text, thanks.